truffle/com.oracle.truffle.api.instrumentation.test/src/com/oracle/truffle/api/instrumentation/InstrumentationTest.java

```diff
@@ -0,0 +1,693 @@
+        Assert.assertEquals(1, MultipleInstrumentsInstrumentation.onDisposeCounter);
+
+        instrument.setEnabled(true);
+        Assert.assertEquals(2, MultipleInstrumentsInstrumentation.constructor);
+        Assert.assertEquals(2, MultipleInstrumentsInstrumentation.onCreateCounter);
+        Assert.assertEquals(1, MultipleInstrumentsInstrumentation.onDisposeCounter);
+
+        instrument.setEnabled(false);
+        Assert.assertEquals(2, MultipleInstrumentsInstrumentation.constructor);
+        Assert.assertEquals(2, MultipleInstrumentsInstrumentation.onCreateCounter);
+        Assert.assertEquals(2, MultipleInstrumentsInstrumentation.onDisposeCounter);
+    }
+
+    @Registration(id = "testMultipleInstruments")
+    public static class MultipleInstrumentsInstrumentation extends TruffleInstrument {
+
+        private static int onCreateCounter = 0;
+        private static int onDisposeCounter = 0;
+        private static int constructor = 0;
+
+        public MultipleInstrumentsInstrumentation() {
+            constructor++;
+        }
+
+        @Override
+        protected void onCreate(Env env, Instrumenter instrumenter) {
+            onCreateCounter++;
+        }
+
+        @Override
+        protected void onDispose(Env env) {
+            onDisposeCounter++;
+        }
+    }
+
+    /*
+     * Test exceptions from language instrumentations are not wrapped into
+     * InstrumentationExceptions. Test that one language cannot instrument another.
+     */
+    @Test
+    public void testLanguageInstrumentationAndExceptions() throws IOException {
+        TestLanguageInstrumentationLanguage.installInstrumentationsCounter = 0;
+        TestLanguageInstrumentationLanguage.createContextCounter = 0;
+        try {
+            engine.eval(Source.fromText("ROOT(EXPRESSION)", null).withMimeType("testLanguageInstrumentation"));
+            Assert.fail("expected exception");
+        } catch (IOException e) {
+            // we assert that MyLanguageException is not wrapped into
+            // InstrumentationException.
+            if (!(e.getCause() instanceof MyLanguageException)) {
+                Assert.fail(String.format("expected MyLanguageException but was %s in %s", e.getCause(), e));
+            }
+        }
+        Assert.assertEquals(1, TestLanguageInstrumentationLanguage.installInstrumentationsCounter);
+        Assert.assertEquals(1, TestLanguageInstrumentationLanguage.createContextCounter);
+
+        // this should run isolated from the language instrumentations.
+        run("STATEMENT");
+    }
+
+    @SuppressWarnings("serial")
+    private static class MyLanguageException extends RuntimeException {
+
+    }
+
+    @TruffleLanguage.Registration(name = "", version = "", mimeType = "testLanguageInstrumentation")
+    public static class TestLanguageInstrumentationLanguage extends TruffleLanguage<Void> implements InstrumentationLanguage<Void> {
+
```

Comment:
Here is my proposed API change to replace InstrumentationLanguage: https://github.com/jtulach/truffle/commit/68c9405c0adb1448541b5e1e0212c001a92439d6